---
title: "Single-cell RNA sequencing of _Phymatopus californicus_ sample PC27 with STARsolo"
author: | 
    | Andrea Elizabeth Acurio Armas, Bulah Wu, Petr Nguyen
    | `r format(Sys.time(), '%B %d, %Y')`
output: github_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(ggrepel)
library(tidyr)
library(dplyr)
library(Seurat)
library(rmarkdown)
library(DropletUtils)
library(patchwork)
library(RColorBrewer)
setwd("/media/nguyen/Data1/github/bulahwoo/pc27/data/02_starsolo/")
```

## STARsolo
Another way to compile the digital expression matrix is using [STARsolo](https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md). We ran STARsolo with the parameters to best match the output of 10x Genomics [CellRanger](https://www.10xgenomics.com/support/software/cell-ranger/latest).

```{bash, engine.opts='-l', echo=FALSE, comment=""}
cat /media/nguyen/Data1/mao/scseq/starsolo/nows26_pc27_8k/sh_starsolo_nows26_pc27_8k.sh
```

Cell filtering is implemented in STARsolo, so the output is a filtered matrix.

```{r matrix}
mtx_pc27 <- ReadSTARsolo("/media/nguyen/Data1/mao/scseq/starsolo/nows26_pc27_8k/Solo.out/Gene/filtered/")
str(mtx_pc27)
```

In total 9341 cells are retained.

## Analysis using Seurat

[Seurat v5](https://satijalab.org/seurat) was used to analyze the data.

First we filtered the data by dropping (1) features detected in less than 3 cells or (2) cells with feature counts less than 200.

```{r prefilter, warning=FALSE, message=FALSE}
so_pc27_pre <- CreateSeuratObject(counts = mtx_pc27, min.cells = 3, min.features = 200, project = "pc27") %>%
               PercentageFeatureSet(pattern = "^agat|^rrn", col.name = "percent.mt")
```

The diagnostic plots below show the total number of unique molecules in each cell (left), the total number of unique genes in each cell (middle), and the percentage of reads mapped to the mitochondrial genome (right).

```{r violin1, echo=FALSE, fig.width=12, fig.height=4}
df_violin_pre <- data.frame(nCount=so_pc27_pre@meta.data$nCount_RNA,
                            nFeature=so_pc27_pre@meta.data$nFeature_RNA,
                            percentMT=so_pc27_pre@meta.data$percent.mt)

pre1 <-
ggplot(df_violin_pre) +
  geom_jitter(aes(x="",y=nCount), position = position_jitter(seed = 1, width = 0.5), size=0.3, color="cornflowerblue", alpha=0.5) +
  geom_violin(aes(x="", y=nCount), fill="orange", color="orange", alpha=0.5) +
  labs(x="PC_27",
       y="Number of molecules detected") +
  #scale_y_continuous(breaks = seq(0,7000,500)) +
  theme_bw() +
  theme(axis.line = element_blank(),
        axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(linewidth = 1, color="black"), aspect.ratio = 1)

pre2 <-
ggplot(df_violin_pre) +
  geom_jitter(aes(x="",y=nFeature), position = position_jitter(seed = 1, width = 0.5), size=0.3, color="cornflowerblue", alpha=0.5) +
  geom_violin(aes(x="", y=nFeature), fill="orange", color="orange", alpha=0.5) +
  labs(x="PC_27",
       y="Number of genes detected") +
  #scale_y_continuous(breaks = seq(0,2400,200)) +
  theme_bw() +
  theme(axis.line = element_blank(),
        axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(linewidth = 1, color="black"), aspect.ratio = 1)

pre3 <-
ggplot(df_violin_pre) +
  geom_jitter(aes(x="",y=percentMT), position = position_jitter(seed = 1, width = 0.5), size=0.3, color="cornflowerblue", alpha=0.5) +
  geom_violin(aes(x="", y=percentMT), fill="orange", color="orange", alpha=0.5) +
  labs(x="PC_27",
       y="Reads mapped to Mt genes (%)") +
  #scale_y_continuous(breaks = seq(0,10,2)) +
  theme_bw() +
  theme(axis.line = element_blank(),
        axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(linewidth = 1, color="black"), aspect.ratio = 1)

pre1 + pre2 + pre3
```

We further filtered cells (1) with feature counts larger than 2500 or less than 200 or (2) with mitochondrial counts larger than 5%.

```{r postfilter, warning=FALSE, message=FALSE}
so_pc27 <- CreateSeuratObject(counts = mtx_pc27, min.cells = 3, min.features = 200, project = "pc27") %>%
           PercentageFeatureSet(pattern = "^agat|^rrn", col.name = "percent.mt") %>%
           subset(subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```

We checked the diagnostic plots after filtering.

```{r violin2, echo=FALSE, fig.width=12, fig.height=4}
df_violin <- data.frame(nCount=so_pc27@meta.data$nCount_RNA,
                        nFeature=so_pc27@meta.data$nFeature_RNA,
                        percentMT=so_pc27@meta.data$percent.mt)

post1 <-
ggplot(df_violin) +
  geom_jitter(aes(x="",y=nCount), position = position_jitter(seed = 1, width = 0.5), size=0.3, color="cornflowerblue", alpha=0.5) +
  geom_violin(aes(x="", y=nCount), fill="orange", color="orange", alpha=0.5) +
  labs(x="PC_27",
       y="Number of molecules detected") +
  #scale_y_continuous(breaks = seq(0,7000,500)) +
  theme_bw() +
  theme(axis.line = element_blank(),
        axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(linewidth = 1, color="black"), aspect.ratio = 1)

post2 <-
ggplot(df_violin) +
  geom_jitter(aes(x="",y=nFeature), position = position_jitter(seed = 1, width = 0.5), size=0.3, color="cornflowerblue", alpha=0.5) +
  geom_violin(aes(x="", y=nFeature), fill="orange", color="orange", alpha=0.5) +
  labs(x="PC_27",
       y="Number of genes detected") +
  #scale_y_continuous(breaks = seq(0,2400,200)) +
  theme_bw() +
  theme(axis.line = element_blank(),
        axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(linewidth = 1, color="black"), aspect.ratio = 1)

post3 <-
ggplot(df_violin) +
  geom_jitter(aes(x="",y=percentMT), position = position_jitter(seed = 1, width = 0.5), size=0.3, color="cornflowerblue", alpha=0.5) +
  geom_violin(aes(x="", y=percentMT), fill="orange", color="orange", alpha=0.5) +
  labs(x="PC_27",
       y="Reads mapped to Mt genes (%)") +
  #scale_y_continuous(breaks = seq(0,10,2)) +
  theme_bw() +
  theme(axis.line = element_blank(),
        axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(linewidth = 1, color="black"), aspect.ratio = 1)

post1 + post2 + post3
```

UMAP clustering of data from 1333 cells revealed 5 cell clusters.

```{r umap, warning=FALSE, message=FALSE}
so_pc27 <- CreateSeuratObject(counts = mtx_pc27, min.cells = 3, min.features = 200, project = "pc27") %>%
           PercentageFeatureSet(pattern = "^agat|^rrn", col.name = "percent.mt") %>%
           subset(subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5) %>%
           SCTransform(vars.to.regress = "percent.mt") %>%
           RunPCA() %>%
           FindNeighbors(dims = 1:30) %>%
           RunUMAP(dims = 1:30) %>%
           FindClusters()
df_umap <- so_pc27@reductions$umap@cell.embeddings %>% as.data.frame() %>% cbind(color=so_pc27@meta.data$seurat_clusters)
#length(unique(df_umap$color))
my_color <- c(brewer.pal(name="Set2", n=8),brewer.pal(name="Dark2", n=8))[c(1,3,2,4,5)]

ggplot(df_umap) +
  geom_point(aes(x=umap_1, y=umap_2, color=color), size=0.8) +
  geom_text_repel(data=df_umap %>% group_by(color) %>% summarise(q1=quantile(umap_1, 0.5), q2=quantile(umap_2, 0.5)),
                  aes(x=q1, y=q2, label = LETTERS[1:5]), size=8) +
  labs(title=expression(italic(P.)~italic(californicus)~"PC27"),
       x="UMAP_1",
       y="UMAP_2") +
  scale_color_manual(values = my_color, name="clusters", labels=LETTERS[1:5]) +
  guides(color = guide_legend(override.aes = list(size = 5))) +
  theme_bw() +
  theme(axis.line = element_blank(),
        axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        legend.title = element_text(size=10),
        legend.background=element_blank(),
        legend.justification=c(1, 0.85),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(linewidth = 1, color="black"), aspect.ratio = 1)
```

Genes differentially expressed among each cluster were selected as marker genes.

```{r marker01, warning=FALSE, message=FALSE}
so_pc27.markers <- FindAllMarkers(so_pc27, only.pos = TRUE, logfc.threshold = 0.25)
so_pc27.markers %>% arrange(desc(avg_log2FC)) %>% group_by(cluster) %>% dplyr::slice(1:3) %>% as.data.frame()
```

The above table shows, in each cluster, three genes with largest average log2 fold-change value (avg_log2FC). We can visualize the marker gene expression (from left to right: top1, top2, top3; from top to bottom: cluster 0, 1, 2, 3 and 4).

```{r marker02, echo=FALSE, fig.width=16, fig.height=4}
default_top1_markers_list <- so_pc27.markers %>% arrange(desc(avg_log2FC)) %>% group_by(cluster) %>% dplyr::slice(1:1) %>% pull(gene)
default_top2_markers_list <- so_pc27.markers %>% arrange(desc(avg_log2FC)) %>% group_by(cluster) %>% dplyr::slice(2:2) %>% pull(gene)
default_top3_markers_list <- so_pc27.markers %>% arrange(desc(avg_log2FC)) %>% group_by(cluster) %>% dplyr::slice(3:3) %>% pull(gene)
dtop1 <- VlnPlot(so_pc27, features=default_top1_markers_list, pt.size=0.2, stack=T, flip=T) + NoLegend()
dtop2 <- VlnPlot(so_pc27, features=default_top2_markers_list, pt.size=0.2, stack=T, flip=T) + NoLegend()
dtop3 <- VlnPlot(so_pc27, features=default_top3_markers_list, pt.size=0.2, stack=T, flip=T) + NoLegend()
dtop1 + dtop2 + dtop3
```

We noticed the high adjusted _P_-value (p_val_adj) for marker genes in all clusters, so we filtered genes with p_val_adj >= 0.05

```{r marker03}
so_pc27.markers %>% filter(p_val_adj < 0.05) %>% arrange(desc(avg_log2FC)) %>% group_by(cluster) %>% dplyr::slice(1:3) %>% as.data.frame()
```

We can visualize the marker gene expression with filtered data. Note that cluster 1 only has one marker gene that meets this filtering criterion.

```{r marker04, echo=FALSE, fig.width=16, fig.height=4}
filtered_top1_markers_list <- so_pc27.markers %>% filter(p_val_adj < 0.05) %>% arrange(desc(avg_log2FC)) %>% group_by(cluster) %>% dplyr::slice(1:1) %>% pull(gene)
filtered_top2_markers_list <- so_pc27.markers %>% filter(p_val_adj < 0.05) %>% arrange(desc(avg_log2FC)) %>% group_by(cluster) %>% dplyr::slice(2:2) %>% pull(gene)
filtered_top3_markers_list <- so_pc27.markers %>% filter(p_val_adj < 0.05) %>% arrange(desc(avg_log2FC)) %>% group_by(cluster) %>% dplyr::slice(3:3) %>% pull(gene)
ftop1 <- VlnPlot(so_pc27, features=filtered_top1_markers_list, pt.size=0.2, stack=T, flip=T) + NoLegend()
ftop2 <- VlnPlot(so_pc27, features=filtered_top2_markers_list, pt.size=0.2, stack=T, flip=T) + NoLegend()
ftop3 <- VlnPlot(so_pc27, features=filtered_top3_markers_list, pt.size=0.2, stack=T, flip=T) + NoLegend()
ftop1 + ftop2 + ftop3
```

The below table shows top10 marker genes in each cluster with inferred gene names derived from annotated reference genomes _Bombyx mori_, _Danaus plexippus_, and _Plutella xylostella_.

```{r marker05}
selected_markers <- so_pc27.markers %>% filter(p_val_adj < 0.05) %>% arrange(desc(avg_log2FC)) %>% group_by(cluster) %>% dplyr::slice(1:10)
gene_name <- read.delim("/media/nguyen/Data1/mao/braker3/phymatopus_californicus/genome_mao/annotation/gene_name/anno_combo_mito", header = F)
selected_markers %>% inner_join(gene_name, by=join_by(gene == V1)) %>% select(!c(V2,V3,V5,V6,V7,V8)) %>% rename(V4="Bombyx_mori") %>% as.data.frame()
```

## Session info

```{r sessioninfo, echo=FALSE}
sessionInfo()
```
